--resume-extract

gperciva opened the pull request.

gperciva commented:
Example, with a weird example of an archive containing two copies of the same file:

```
$ ./tarsnap -tf random-100k-random-100k
random-100k
random-100k
$ rm random-100k
$ ./tarsnap --debug-network-stats -xf random-100k-random-100k
Connection ended with in / out:	204029	1185
$ rm random-100k
$ ./tarsnap --debug-network-stats -xf random-100k-random-100k --resume-extract
Connection ended with in / out:	103500	1039
```

cperciva commented:
Do we want to support `resume-extract` as a configuration file setting?  It seems like it could be dangerous as a "set and forget" option -- unless there's a compelling reason it seems like it might be better to leave it as command-line-only until someone comes forward saying that they need to put it into a configuration file.

And if we're not supporting it as a configuration file option, we don't need `--no-resume-extract`.

cperciva commented:
While I agree that `--resume-extract` should *normally* be safe, I can imagine situations where it could cause problems; for example, if someone archives (using `@-` functionality) synthetic tarballs created by tools (e.g., database dumps) which don't fill in timestamps.  If someone deliberately uses `--resume-extract` when extracting such broken archives they deserve what they get; but I could imagine someone setting that in a configuration file and forgetting about it.

So for the sake of minimizing user error I think let's keep it as command-line-only... at least until someone steps forward and says "I really wish I could put this into my configuration file because <insert use case here>".

gperciva commented:
Ok, removed `resume-extract` from a config file option.  Rebased history.

cperciva commented on tar/bsdtar.c:
```diff
@@ -1779,6 +1782,11 @@ dooption(struct bsdtar *bsdtar, const char * conf_opt,
 
 		bsdtar->option_quiet = 1;
 		bsdtar->option_quiet_set = 1;
+	} else if (strcmp(conf_opt, "resume-extract") == 0) {
```
This code here is for processing configuration file options.  Why do we need it?

cperciva commented on tar/bsdtar.c:
```diff
@@ -712,6 +712,9 @@ main(int argc, char **argv)
 		case OPTION_RECOVER:
 			set_mode(bsdtar, opt, "--recover");
 			break;
+		case OPTION_RESUME_EXTRACT:
```
Why not just set `bsdtar->option_resume_extract` here?

cperciva commented on tar/tarsnap.1-man.in:
```diff
@@ -624,6 +624,11 @@ Ignore any
 \fBquiet\fP
 option specified in a configuration file.
 .TP
+\fB\--no-resume-extract\fP
```
Not needed any more?

cperciva commented on tar/tarsnap.1-mdoc.in:
```diff
@@ -561,6 +561,10 @@ option specified in a configuration file.
 Ignore any
 .Cm quiet
 option specified in a configuration file.
+.It Fl -no-resume-extract
```
Not needed any more?

cperciva commented:
Looks like there's still some `--no-resume-extract` scattered around.

gperciva commented:
Sorry, was distracted by a viola gig.

- `git grep no-resume` now returns nothing
- I added `(x mode only)` to the man-page
- fixed a minor typo in NEWS ("an --dump-config" -> "a --dump-config")

cperciva commented on tar/read.c:
```diff
@@ -265,6 +299,20 @@ read_archive(struct bsdtar *bsdtar, char mode)
+			/* Don't extract if the file already matches it. */
+			if (bsdtar->option_resume_extract) {
+				if ((r = check_skip_file(
+				    archive_entry_pathname(entry), st)) == -1) {
```
How carefully did you check that `archive_entry_pathname(entry)` is the correct path to the file we would create?  I *think* this is correct but with options like `-C` and `-s` and the handling of absolute paths and paths containing `..` I want to be sure that we have this right and aren't accidentally comparing metadata against the wrong place.

gperciva commented:
I think this is working.  I made sure that `check_skip_file()` came after `edit_pathname()`.  See below for the tests; I think I covered all the weird cases.

## Basic usage:

```
$ ./tarsnap -tf random-100k-random-100k
random-100k
random-100k
$ rm -f random-100k
$ ./tarsnap --debug-network-stats -xf random-100k-random-100k
Connection ended with in / out:	204029	1185
$ rm -f random-100k
$ ./tarsnap --debug-network-stats -xf random-100k-random-100k --resume-extract
Connection ended with in / out:	103500	1039
$ ./tarsnap --debug-network-stats -xf random-100k-random-100k --resume-extract
Connection ended with in / out:	2483	747
$
```

## -C out/

```
$ mkdir out/
$ ./tarsnap --debug-network-stats -xf random-100k-random-100k --resume-extract -C out/
Connection ended with in / out:	103500	1039
$ 
```

## -s

```
$ rm -f foo-100k
$ ./tarsnap --debug-network-stats -xf random-100k-random-100k --resume-extract -s "/random/foo/g"
Connection ended with in / out:	103500	1039
$ ./tarsnap --debug-network-stats -xf random-100k-random-100k --resume-extract -s "/random/foo/g"
Connection ended with in / out:	2483	747
$ 
```

## -C out/ -s

```
$ mkdir out/
$ ./tarsnap --debug-network-stats -xf random-100k-random-100k --resume-extract -s "/random/foo/g" -C out
Connection ended with in / out:	103500	1039
$ ls out/
foo-100k
$
```

## absolute paths combined with -C

```
$ ./tarsnap -tf random-100k-random-100k-absolute
/home/td/src/tarsnap/build/random-100k
/home/td/src/tarsnap/build/random-100k
$ rm -f random-100k
$ ./tarsnap --debug-network-stats -xf random-100k-random-100k-absolute --resume-extract -P -C out
Connection ended with in / out:	103598	1039
$ ./tarsnap --debug-network-stats -xf random-100k-random-100k-absolute --resume-extract -P -C out
Connection ended with in / out:	2581	747
td@mac: ~/src/tarsnap/build (resume-extract)
$ ls out/
td@mac: ~/src/tarsnap/build (resume-extract)
$ 
```

## relative dirs

```
$ cd out/
$ ls
$ ../tarsnap --debug-network-stats -x -f random-100k-random-100k-updir --resume-extract -P
Connection ended with in / out:	103524	1039
$ ../tarsnap --debug-network-stats -x -f random-100k-random-100k-updir --resume-extract -P
Connection ended with in / out:	2507	747
$ ls
$ 
```

cperciva commented on tar/tarsnap.1-man.in:
```diff
@@ -742,6 +742,11 @@ but it is likely that other
+(x mode only)
+Don't extract files whose filesize and mtime matches existing files on the
+disk.  Primarily used to resume an archive extraction which was interrupted.
+.TP
```
Probably worth adding something along the lines of "This differs from -k in that --resume-extract will overwrite a file if the size or modification time do not match, as can happen if tarsnap is killed partway through extracting a file."

cperciva commented on NEWS.md:
```diff
@@ -7,6 +7,8 @@
   non-blank lines read from config files.
 - tarsnap now gives an error if there are unused command-line arguments.
   (i.e. "tarsnap -d -f a1 a2", where "a2" is unused.)
+- tarsnap now accepts a --resume-extract option to skip extracting files whose
```
This needs to have a merge conflict fixed.

cperciva commented on tar/read.c:
```diff
@@ -111,6 +111,40 @@ progress_func(void * cookie)
+		goto err0;
+	}
+
+	/* Compare file size and mtime. */
```
The comment here should point out that we're comparing the *seconds* of the mtime, not the complete timespec; some filesystems don't have sub-second timestamp precision, so comparing timespecs would produce a lot of false negatives.

cperciva commented:
Ok, looks good.  I've confirmed by inspection of the libarchive source code that archive_entry_pathname is definitely the path to where we're going to write on disk.

gperciva commented:
Rebased to fix merge conflicts, but I left the two recent fixes to your comments are separate commits for easy examination.

Fun fact: I misread the POSIX and linux man-pages about `stat(2)`.  My brain just couldn't process that it was actually `st_mtim` instead of `st_mtime`.  The latter is `#define`d as `st_mtim.tv_sec` for backwards compatibility on my desktop, so I didn't realize the discrepancy.

gperciva commented:
Updated to current master.  I checked that all the tests in https://github.com/Tarsnap/tarsnap/pull/327#issuecomment-438416561 work.
